File: src/core/cleaner.ts
import type { CleaningResult, Report, CleanOptions } from '../types/index.js';
import { getModuleByIds, type Module } from '../modules/index.js';
import { logger } from '../utils/logger.js';

export class Cleaner {
  private modules: Module[];
  private dryRun: boolean;
  private force: boolean;

  constructor(moduleIds: string[], options: CleanOptions) {
    this.modules = getModuleByIds(moduleIds);
    this.dryRun = options.dryRun;
    this.force = options.force;
  }

  async clean(): Promise<Report> {
    const startTime = new Date();
    const results: CleaningResult[] = [];
    const errors: string[] = [];

    if (this.dryRun) {
      logger.warn('Modo DRY-RUN ativo - nenhuma alteração será feita');
    }

    for (const module of this.modules) {
      if (!module.isAvailable()) {
        logger.debug(`${module.name} não disponível, pulando`);
        continue;
      }

      logger.startSpinner(`Limpando ${module.name}...`);
      try {
        const result = await module.clean(this.dryRun, this.force);
        results.push(result);

        if (result.success) {
          logger.stopSpinner(true, `${module.name}: ${logger.formatBytes(result.spaceFreed)} liberada`);
        } else {
          logger.stopSpinner(false, `${module.name}: Erro`);
        }

        if (result.errors.length > 0) {
          errors.push(...result.errors);
        }
      } catch (error) {
        logger.stopSpinner(false, `Erro ao limpar ${module.name}`);
        errors.push(`${module.name}: ${(error as Error).message}`);
        results.push({
          module: module.id,
          success: false,
          spaceFreed: 0,
          itemsRemoved: 0,
          errors: [(error as Error).message],
        });
      }
    }

    const endTime = new Date();

    return {
      startTime,
      endTime,
      modules: results,
      totalSpaceFreed: results.reduce((sum, r) => sum + r.spaceFreed, 0),
      totalItemsRemoved: results.reduce((sum, r) => sum + r.itemsRemoved, 0),
      errors,
    };
  }

  printReport(report: Report): void {
    logger.title('Relatório de Limpeza');

    const duration = report.endTime.getTime() - report.startTime.getTime();
    const durationStr = duration < 60000 
      ? `${Math.round(duration / 1000)}s`
      : `${Math.round(duration / 60000)}min`;

    logger.info(`Tempo de execução: ${durationStr}`);
    logger.space();

    if (report.modules.length > 0) {
      logger.subtitle('Módulos processados:');
      for (const result of report.modules) {
        if (result.success) {
          logger.item(`${result.module}: ${logger.formatBytes(result.spaceFreed)}`, `${result.itemsRemoved} itens`);
        } else {
          logger.item(`${result.module}: ERRO`, 'Falhou');
        }
      }
    }

    logger.space();
    logger.subtitle('Resumo:');
    logger.info(`Espaço liberado: ${logger.formatBytes(report.totalSpaceFreed)}`);
    logger.info(`Itens removidos: ${report.totalItemsRemoved}`);
    logger.info(`Módulos com erro: ${report.errors.length}`);

    if (report.errors.length > 0) {
      logger.space();
      logger.warn('Erros encontrados:');
      for (const error of report.errors) {
        logger.item(error);
      }
    }

    logger.space();
    if (this.dryRun) {
      logger.success('Dry-run concluído - use sem --dry-run para aplicar');
    } else {
      logger.success('Limpeza concluída com sucesso!');
    }
  }
}

export function createCleaner(moduleIds: string[], options: CleanOptions): Cleaner {
  return new Cleaner(moduleIds, options);
}

File: internal/core/cleaner.go
package core

import (
	"fmt"
	"strings"
	"time"

	"github.com/joaomjbraga/piunter/internal/modules"
	"github.com/joaomjbraga/piunter/internal/utils"
	"github.com/joaomjbraga/piunter/pkg/types"
)

type Cleaner struct {
	modules []modules.Module
	dryRun  bool
	verbose bool
}

func NewCleaner(moduleIds []string, dryRun bool) *Cleaner {
	mods := modules.GetModulesByIds(moduleIds)
	return &Cleaner{
		modules: mods,
		dryRun:  dryRun,
		verbose: false,
	}
}

func NewCleanerWithOptions(moduleIds []string, dryRun bool, verbose bool) *Cleaner {
	mods := modules.GetModulesByIds(moduleIds)
	return &Cleaner{
		modules: mods,
		dryRun:  dryRun,
		verbose: verbose,
	}
}

func (c *Cleaner) Clean() (*types.Report, error) {
	startTime := time.Now()
	results := c.cleanSequential()
	endTime := time.Now()

	return buildReport(startTime, endTime, results), nil
}

func isVerboseEnabled(verboseFlag bool, configVerbose bool) bool {
	return verboseFlag || configVerbose
}

func BuildConfirmationSummary(moduleIDs []string, dryRun bool) string {
	if len(moduleIDs) == 0 {
		return "Nenhum módulo selecionado."
	}

	base := fmt.Sprintf("Os seguintes módulos serão processados: %s.", strings.Join(moduleIDs, ", "))
	if dryRun {
		return base + " Execução em modo dry-run, nenhuma alteração será aplicada."
	}
	return base + " Confirmação necessária para aplicar as alterações."
}

func buildReport(startTime, endTime time.Time, results []types.CleaningResult) *types.Report {
	var totalSpaceFreed int64
	var totalItemsRemoved int
	var errors []string

	for _, r := range results {
		totalSpaceFreed += r.SpaceFreed
		totalItemsRemoved += r.ItemsRemoved
		errors = append(errors, r.Errors...)
	}

	return &types.Report{
		StartTime:         startTime.Format(time.RFC3339),
		EndTime:           endTime.Format(time.RFC3339),
		Modules:           results,
		TotalSpaceFreed:   totalSpaceFreed,
		TotalItemsRemoved: totalItemsRemoved,
		Errors:            errors,
	}
}

func (c *Cleaner) cleanSequential() []types.CleaningResult {
	var results []types.CleaningResult

	for _, m := range c.modules {
		if !m.IsAvailable() {
			continue
		}

		result, err := m.Clean(c.dryRun)
		if c.verbose {
			utils.Info(fmt.Sprintf("[verbose] módulo %s concluído", m.ID()))
		}
		if err != nil {
			results = append(results, types.CleaningResult{
				Module:       m.ID(),
				Success:      false,
				SpaceFreed:   0,
				ItemsRemoved: 0,
				Errors:       []string{fmt.Sprintf("%s: %s", m.Name(), err.Error())},
			})
			continue
		}
		results = append(results, *result)
	}

	return results
}

func summarizeReport(report *types.Report) string {
	if len(report.Errors) == 0 {
		return fmt.Sprintf("Limpeza concluída com sucesso. Espaço liberado: %s", utils.FormatBytes(report.TotalSpaceFreed))
	}
	return fmt.Sprintf("Limpeza concluída com avisos. Espaço liberado: %s. Erros: %d", utils.FormatBytes(report.TotalSpaceFreed), len(report.Errors))
}

func (c *Cleaner) PrintReport(report *types.Report) {
	startTime, err := time.Parse(time.RFC3339, report.StartTime)
	if err != nil {
		startTime = time.Now()
	}
	endTime, err := time.Parse(time.RFC3339, report.EndTime)
	if err != nil {
		endTime = time.Now()
	}
	duration := endTime.Sub(startTime)

	var durationStr string
	if duration.Seconds() < 60 {
		durationStr = fmt.Sprintf("%.1fs", duration.Seconds())
	} else {
		durationStr = fmt.Sprintf("%.1fmin", duration.Minutes())
	}

	for _, r := range report.Modules {
		if r.Success {
			utils.Item(r.Module, utils.FormatBytes(r.SpaceFreed))
		} else {
			utils.Item(r.Module, "\033[31merro\033[0m")
		}
	}

	utils.Space()
	fmt.Printf("  \033[90m%s\033[0m\n", strings.Repeat("─", 40))

	totalSize := utils.FormatBytes(report.TotalSpaceFreed)
	totalItems := fmt.Sprintf("%d", report.TotalItemsRemoved)
	totalErrors := fmt.Sprintf("%d", len(report.Errors))

	fmt.Println()
	fmt.Printf("  \033[1mResumo\033[0m\n")
	fmt.Printf("    \033[90m-\033[0m \033[37mEspaço liberado:\033[0m \033[32m%s\033[0m\n", totalSize)
	fmt.Printf("    \033[90m-\033[0m \033[37mItens removidos:\033[0m \033[36m%s\033[0m\n", totalItems)

	errorColor := "\033[32m"
	if len(report.Errors) > 0 {
		errorColor = "\033[31m"
	}
	fmt.Printf("    \033[90m-\033[0m \033[37mErros:\033[0m %s%s\033[0m\n", errorColor, totalErrors)

	if len(report.Errors) > 0 {
		utils.Space()
		fmt.Printf("  \033[33mAvisos:\033[0m\n")
		for _, err := range report.Errors {
			fmt.Printf("    \033[90m-\033[0m \033[33m%s\033[0m\n", err)
		}
	}

	utils.Space()

	fmt.Printf("  \033[32m*\033[0m %s\n", summarizeReport(report))
	if c.dryRun {
		fmt.Printf("  \033[90mExecute sem --dry-run para aplicar\033[0m\n")
	} else {
		fmt.Printf("  \033[90mTempo total: %s\033[0m\n", durationStr)
	}
	fmt.Println()
}